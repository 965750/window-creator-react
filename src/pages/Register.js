import { connect } from 'react-redux'
import { injectIntl } from 'react-intl'
import { Redirect } from 'react-router-dom'
import { register } from '../store/actions/authActions'

import BaseInput from '../components/inputs/BaseInput'
import BaseSubmit from '../components/inputs/BaseSubmit'
import PasswordStrength from '../components/blocks/PasswordStrength'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './Register.module.scss'

class Register extends Component {
  state = {
    email: '',
    password: '',
    passwordRepeat: '',
    firstName: '',
    lastName: '',
    error: null,
  }

  handleChange = e => {
    this.setState({
      [e.target.id]: e.target.value,
    })
  }

  onRegister = e => {
    e.preventDefault()

    this.props.register({
      email: this.state.email,
      password: this.state.password,
      firstName: this.state.firstName,
      lastName: this.state.lastName,
    })
  }

  validationError = error => {
    this.setState({
      error,
    })
  }

  render() {
    if (this.props.isLoggedIn) return <Redirect to="/creator" />

    return (
      <form
        className={`${styles.form__wrapper} mt-10 mx-auto`}
        onSubmit={this.onRegister}
      >
        <BaseInput
          value={this.state.email}
          validationError={this.validationError}
          handleChange={this.handleChange}
          id="email"
          type="text"
          toValidate={['email', 'isRequired']}
          classes="mb-3"
          placeholder="Email address"
        />
        <PasswordStrength password={this.state.password} />
        <div className="flex flex-wrap">
          <BaseInput
            value={this.state.password}
            validationError={this.validationError}
            handleChange={this.handleChange}
            id="password"
            type="password"
            toValidate={['isRequired', 'passwordLength']}
            classes="mb-3 w-4/9"
            placeholder="Password"
          />
          <BaseInput
            value={this.state.passwordRepeat}
            validationError={this.validationError}
            handleChange={this.handleChange}
            id="passwordRepeat"
            type="password"
            toValidate={['isRequired', 'match']}
            valueToMatch={this.state.password}
            classes="mb-3 ml-auto w-4/9"
            placeholder="Repeat password"
          />
        </div>
        <BaseInput
          value={this.state.firstName}
          validationError={this.validationError}
          handleChange={this.handleChange}
          id="firstName"
          type="text"
          toValidate={['isRequired']}
          classes="mb-3"
          placeholder="First name"
        />
        <BaseInput
          value={this.state.lastName}
          validationError={this.validationError}
          handleChange={this.handleChange}
          id="lastName"
          type="text"
          toValidate={['isRequired']}
          classes="mb-3"
          placeholder="Last name"
        />
        <BaseSubmit
          value="Register"
          disabled={
            this.state.error ||
            !this.state.email ||
            !this.state.password ||
            !this.state.passwordRepeat ||
            !this.state.firstName ||
            !this.state.lastName
          }
        />
      </form>
    )
  }
}

const mapStateToProps = state => ({
  isLoggedIn: state.firebase.auth.uid,
})

const mapDispatchToProps = dispatch => ({
  register: credentials => dispatch(register(credentials)),
})

Register.propTypes = {
  isLoggedIn: PropTypes.string,
  register: PropTypes.func,
}

export default injectIntl(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Register)
)
