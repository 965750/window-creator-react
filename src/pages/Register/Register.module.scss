.form {

  &__wrapper {
    width: 400px;

    @media screen and (max-width: 640px) {
      width: 300px;
    }
  }
}