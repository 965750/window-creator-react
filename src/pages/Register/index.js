import { connect } from 'react-redux'
import { injectIntl } from 'react-intl'
import { Redirect } from 'react-router-dom'
import { register } from '../../store/actions/authActions'

import BaseInput from '../../components/inputs/BaseInput'
import BaseSubmit from '../../components/inputs/BaseSubmit'
import PasswordStrength from '../../components/blocks/PasswordStrength'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './Register.module.scss'

class Register extends Component {
  state = {
    email: '',
    password: '',
    passwordRepeat: '',
    firstName: '',
    lastName: '',
    formError: true,
    error: {
      email: null,
      password: null,
      passwordRepeat: null,
      firstName: null,
      lastName: null,
    },
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.error !== prevState.error) {
      this.checkFormValidation()
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.id]: e.target.value,
    })
  }

  checkFormValidation = () => {
    if (!this.state.error.email && !this.state.error.password && !this.state.error.passwordRepeat && !this.state.error.firstName && !this.state.error.email && !this.state.error.lastName) {
      this.setState({
        formError: null,
      })
    } else {
      this.setState({
        formError: true,
      })
    }
  }

  onRegister = (e) => {
    e.preventDefault()

    this.props.register({
      email: this.state.email,
      password: this.state.password,
      firstName: this.state.firstName,
      lastName: this.state.lastName,
    })
  }

  validationError = (error, id) => {
    const newError = { ...this.state.error } // eslint-disable-line
    newError[id] = error

    this.setState({
      error: newError,
    })
  }

  render() {
    if (this.props.isLoggedIn) return <Redirect to="/creator" />

    return (
      <form
        className={`${styles.form__wrapper} mt-10 mx-auto`}
        onSubmit={this.onRegister}
      >
        <BaseInput
          value={this.state.email}
          validationError={this.validationError}
          handleChange={this.handleChange}
          id="email"
          type="text"
          toValidate={['email', 'isRequired']}
          classes="mb-3"
          placeholder="Email address"
        />
        <PasswordStrength password={this.state.password} />
        <div className="flex flex-wrap">
          <BaseInput
            value={this.state.password}
            validationError={this.validationError}
            handleChange={this.handleChange}
            id="password"
            type="password"
            toValidate={['isRequired', 'passwordLength']}
            classes="mb-3 w-4/9"
            placeholder="Password"
          />
          <BaseInput
            value={this.state.passwordRepeat}
            validationError={this.validationError}
            handleChange={this.handleChange}
            id="passwordRepeat"
            type="password"
            toValidate={['isRequired', 'match']}
            valueToMatch={this.state.password}
            classes="mb-3 ml-auto w-4/9"
            placeholder="Repeat password"
          />
        </div>
        <BaseInput
          value={this.state.firstName}
          validationError={this.validationError}
          handleChange={this.handleChange}
          id="firstName"
          type="text"
          toValidate={['isRequired']}
          classes="mb-3"
          placeholder="First name"
        />
        <BaseInput
          value={this.state.lastName}
          validationError={this.validationError}
          handleChange={this.handleChange}
          id="lastName"
          type="text"
          toValidate={['isRequired']}
          classes="mb-3"
          placeholder="Last name"
        />
        <BaseSubmit
          classes="mx-auto"
          value="Register"
          disabled={
            this.state.formError
            || !this.state.email
            || !this.state.password
            || !this.state.passwordRepeat
            || !this.state.firstName
            || !this.state.lastName
          }
        />
      </form>
    )
  }
}

const mapStateToProps = (state) => ({
  isLoggedIn: state.firebase.auth.uid,
})

const mapDispatchToProps = (dispatch) => ({
  register: (credentials) => dispatch(register(credentials)),
})

Register.propTypes = {
  isLoggedIn: PropTypes.string,
  register: PropTypes.func,
}

export default injectIntl(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Register),
)
