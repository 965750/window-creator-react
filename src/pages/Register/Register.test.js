import '@testing-library/jest-dom/extend-expect'

import { BrowserRouter } from 'react-router-dom'
import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from '@testing-library/react';

import Register from './index'
import messages from '../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
    stepper: {
      active: 1,
      steps: [
        {
          label: 'Step 1',
          title: 'Choose Door',
          id: 1,
        },
        {
          label: 'Step 2',
          title: 'Choose Door Division',
          id: 2,
        },
        {
          label: 'Step 3',
          title: 'Choose Color',
          id: 3,
        },
      ],
    }
  },
  creator: {
    window: {
      color: '#000',
      width: 110,
      height: 230,
      doorType: 2,
      rows: [0, 1, 2],
      columns: [0, 1, 2],
    },
  },
  firebase: {
    auth: {
      uid: null,
    }
  }
}

const reducer = (state = initialStore, action) => {
  switch (action.type) {
    case 'MOCK_LOGIN':
      return {
        ...state,
        firebase: {
          auth: {
            uid: 'mockUid',
          }
        }
      }
    case 'RESET_STORE':
      return initialStore
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('Register', () => {
  let wrapper
  const initialProps = {}

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <BrowserRouter>
        <IntlProvider
          locale="en"
          messages={messages[store.getState().theme.lang]}
        >
          <Register {...props} />
        </IntlProvider>
      </BrowserRouter>,
    )
  }

  afterEach(() => {
    cleanup

    store.dispatch({
      type: 'RESET_STORE',
    })
  })

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should redirect when user is logged in', () => {
    setUp()

    store.dispatch({
      type: 'MOCK_LOGIN',
    })

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should enable submit button after correctly filling form', () => {
    setUp()

    expect(wrapper.getByTestId('submitWrapper').disabled).toBe(true)

    fireEvent.change(wrapper.getByPlaceholderText('Adres Email'), { target: { value: 'mockedEmail@gmail.com' }})
    fireEvent.change(wrapper.getByPlaceholderText('Hasło'), { target: { value: 'mockPassword' }})
    fireEvent.change(wrapper.getByPlaceholderText('Powtórz hasło'), { target: { value: 'mockPassword' }})
    fireEvent.change(wrapper.getByPlaceholderText('Imie'), { target: { value: 'mockName' }})
    fireEvent.change(wrapper.getByPlaceholderText('Nazwisko'), { target: { value: 'mockLastName' }})

    expect(wrapper.getByTestId('submitWrapper').disabled).toBe(false)
  })
})
