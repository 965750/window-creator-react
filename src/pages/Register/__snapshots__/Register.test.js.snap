// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="form__wrapper mt-10 mx-auto"
  >
    <div
      class="field__wrapper mb-3"
      data-testid="inputWrapper"
    >
      <input
        class="field  border border-silver rounded p-3 w-full"
        id="email"
        placeholder="Adres Email"
        type="text"
        value=""
      />
    </div>
    <div
      class="bar relative w-4/9 mb-1"
    >
      <p
        class="bar__label top-10- text-osloGray"
      />
      <div
        class="bar__back rounded"
      />
      <div
        class="bar__cover rounded-r absolute bg-white opacity-75 h-full bottom-0 right-0 w-50"
        data-testid="passwordStrength"
        style="width: 100%;"
      />
    </div>
    <div
      class="flex flex-wrap"
    >
      <div
        class="field__wrapper mb-3 w-4/9"
        data-testid="inputWrapper"
      >
        <input
          class="field  border border-silver rounded p-3 w-full"
          id="password"
          placeholder="Hasło"
          type="password"
          value=""
        />
      </div>
      <div
        class="field__wrapper mb-3 ml-auto w-4/9"
        data-testid="inputWrapper"
      >
        <input
          class="field  border border-silver rounded p-3 w-full"
          id="passwordRepeat"
          placeholder="Powtórz hasło"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field__wrapper mb-3"
      data-testid="inputWrapper"
    >
      <input
        class="field  border border-silver rounded p-3 w-full"
        id="firstName"
        placeholder="Imie"
        type="text"
        value=""
      />
    </div>
    <div
      class="field__wrapper mb-3"
      data-testid="inputWrapper"
    >
      <input
        class="field  border border-silver rounded p-3 w-full"
        id="lastName"
        placeholder="Nazwisko"
        type="text"
        value=""
      />
    </div>
    <input
      class="submit  mx-auto cursor-pointer block bg-mountainMeadow rounded text-white"
      data-testid="submitWrapper"
      disabled=""
      type="submit"
      value="Zarejestruj"
    />
  </form>
</DocumentFragment>
`;

exports[`Register should redirect when user is logged in 1`] = `<DocumentFragment />`;
